### PR TITLE
[Doc] Update documentation for no-transitive-change

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
(Some backgrounds, not required to read: https://discourse.llvm.org/t/rfc-c-20-modules-introduce-thin-bmi-and-decls-hash/74755)

This is the document part for the no-transitive-change (https://github.com/llvm/llvm-project/pull/86912, https://github.com/llvm/llvm-project/pull/92083, https://github.com/llvm/llvm-project/pull/92085, https://github.com/llvm/llvm-project/pull/92511) to provide the ability for build system to skip some unnecessary recompilations.

See the patch for examples.

CC: @mathstuf @boris-kolpackov

### CODE REVIEW

**Comment by ilya-biryukov** [comment]
I really like the idea of documenting this, hopefully this create some fruitful discussions with the build system developers and other compiler users.

After trying to put myself into the shoes of the compiler users reading this, I can't help but think that the "no transitive changes"  is slightly confusing and hard to grasp. The example does help quite a bit, but I also feel there should be a way to document this in shorter form. (E.g. we can double-down on the "this is a build optimization" angle and being explicit that we want people to try this and let them know how to report bugs about it?)

I wanted to take a little bit of time to think how to do it before providing concrete suggestions, but still wanted to mention this.

**Comment by ChuanqiXu9** [comment]
> I really like the idea of documenting this, hopefully this create some fruitful discussions with the build system developers and other compiler users.
> 
> After trying to put myself into the shoes of the compiler users reading this, I can't help but think that the "no transitive changes" is slightly confusing and hard to grasp. The example does help quite a bit, but I also feel there should be a way to document this in shorter form. (E.g. we can double-down on the "this is a build optimization" angle and being explicit that we want people to try this and let them know how to report bugs about it?)
> 
> I wanted to take a little bit of time to think how to do it before providing concrete suggestions, but still wanted to mention this.

How about adding a sentence like "this section is primarily for build system vendors. For end compiler users, if you don't want to read it all, this is helpful to reduce recompilations. We encourage build system vendors and end users try this out and bring feedbacks"?

File: clang/docs/ReleaseNotes.rst
Comment (ilya-biryukov) [inline clang/docs/ReleaseNotes.rst:232]: Maybe explicitly mention the user-facing effect of what "no transitive change" means and mention that we want feedback on this?

Something like:
- Clang implemented improvements to BMI of C++20 Modules that should reduce the number of rebuilds during incremental recompilation. We are seeking feedback from Build System authors and other interested users, especially when you feel Clang changes the BMI and missses an opportunity to avoid recompilations or causes correctness issues. See `StandardCPlusPlusModules <StandardCPlusPlusModules.html>`_ for more details. 

Also, maybe put this at the top to increase chances of people noticing this? (if we do want people filing bugs about it, it'd help to increase the chances that they do).

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [inline clang/docs/StandardCPlusPlusModules.rst:655]: Is it actually experimental if it's the default and there is no way to disable it?
I suggest saying "initial", would it better to capture what we intend to do (I suspect we want to keep improving this and want feedback where it falls short)?

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [inline clang/docs/StandardCPlusPlusModules.rst:658]: NIT: `Clang 19` ? I believe we start the name with an uppercase letter everywhere.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [inline clang/docs/StandardCPlusPlusModules.rst:659]: NIT: Maybe call it "Avoiding unnecessary recompilations"?
It should carry the information people are interested in, but is more grounded in the terms that users are likely to recognize. We could introduce what it does right away:


```
Before Clang 19, a change in BMI of any (transitive) dependency would case the outputs of the BMI to change. Starting with Clang 19, changes to non-direct dependencies should not directly affect the output BMI, unless they affect the results of the compilations.
We expect that there are many more opportunities for this optimization than we currently have realized and would appreaciate feedback about missed optimization opportunities.
... 
```

Having an example after that should clarify this even more.
Does that seems reasonable? (Sorry if I got something wrong, I've probably speculated a bit beyond my undestanding in the comment I wrote)

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:722]:
```suggestion
it becomes possible to skip the recompilation of ``Use.cc``
because the altered interfaces do not affect ``Use.cc`` in any way;
there are no transitive changes.
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:726]:
```suggestion
When clang generates a BMI, it records the hash values of all potentially contributory BMIs
for the BMI being produced. This ensures that build systems are not required to consider
transitively imported modules when deciding whether to recompile.
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:737]:
```suggestion
What is considered to be a potential contributory BMIs is currently unspecified. However, it is a severe bug for a BMI to remain unchanged following an observable change that affects its consumers.

We recommend that build systems support this feature as a configurable option so that users can go back to the transitive change mode safely at any time.
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [reply clang/docs/StandardCPlusPlusModules.rst:655]: I'd like to find a better term than "no transitive change" -- this doesn't work well as a noun phrase in English, so we say awkward things like "...the no transitive change feature can be more powerful if ..." I think what bothers me about is the "no" and "change" part -- "change" is a verb, so it comes out sounding verb-like and we end up tacking on "feature" to try to make it a noun again.

That said, I don't have a significantly better phrase currently. :-)

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:743]: I'm not certain what "if the change can be reduced" is trying to say. Is it trying to say that removing code will have less transitive impacts on other BMIs than adding code?

**Review by AaronBallman — COMMENTED** [review]
Some wording changes; more may come as I better understand the concept.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (h-vetinari) [reply clang/docs/StandardCPlusPlusModules.rst:655]: * "no-cascade" feature?
* "isolated BMI" (not sure if this is a great idea given how there's already full/reduced BMI)
* "BMI isolation"
* + variations with synonyms for isolated/secluded/confined/...

File: clang/docs/ReleaseNotes.rst
Comment (ChuanqiXu9) [reply clang/docs/ReleaseNotes.rst]: Good suggestions. Applied. (Just change `should` to `can`. Since I feel `should` read like users can get this benefit directly. But it is not the case. They need helps from build systems).

And I put this into the `What's New in Clang |release|?` section, which is highest except the `Potentially Breaking Change` section.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done. The suggestion looks good. What do you mean for `an example`? There are a following example. Do you mean a example that makes the BMI change but actually it can avoid that?

An example may be:

```
// A.cppm
export module A;
export int a();
```

```
// B.cppm
export module B;
import A;
export inline int b() { return a(); }
```

and we add a unrelated function to A:

```
// A.cppm
export module A;
export int a0();
export int a();
```

Now the users may think the change is unrelated to B and the BMI of B should keep unchanged. But we can't do that since the used DeclID for `a` in module B become to `<1, 1>` instead of `<1, 0>` since the new declaration changes the DeclID. 

Is this the example you want? I didn't post this since I feel it has too many implementation details.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: For example,

```
// B.cppm
export module B;
import A;
export int b() { return a(); }
```

If reduced BMI is not enabled, we will record `a()` part in the BMI of `B`, then we will think module A contributes to module B. Then the BMI of `B` should change if module A changes like the example shows.

However, if reduced BMI enabled, we won't record the function body for `int b()`, then module doesn't contributes to module B. So that we can avoid the changes for B if module A changes like the example shows.

Maybe it will be better to remove the `if the change can be reduced` part. I'll add a sentence to describe the above idea to make it clear.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (h-vetinari) [reply clang/docs/StandardCPlusPlusModules.rst:655]: A slightly cleaner formulation might be "non-cascading X" or perhaps "non-percolating Y".

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [inline clang/docs/StandardCPlusPlusModules.rst:662]:
```suggestion
Before Clang 19, a change in BMI of any (transitive) dependency would cause the
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [inline clang/docs/StandardCPlusPlusModules.rst:732]:
```suggestion
When Clang generates a BMI, it records the hash values of all potentially contributory BMIs
```

or

```suggestion
When ``clang`` generates a BMI, it records the hash values of all potentially contributory BMIs
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [inline clang/docs/StandardCPlusPlusModules.rst:780]:
```suggestion
  $ md5sum B.pcm
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [inline clang/docs/StandardCPlusPlusModules.rst:763]:
```suggestion
  $ md5sum B.pcm
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [inline clang/docs/StandardCPlusPlusModules.rst:741]: How does this happen? Is something missing in the example commands? They look "normal" to me.

Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst:741]: Do you mean in what cases the users need to go back to normal mode? The answer may be when they meet compiler bugs.

Or if you're asking about the command line options, there is no such option in the compiler side. We refactored the format in the BMI so it is in some level always "enabled". But the users have to wait the build system's support to feel so called "no-transitive-change". So the "option" here means a build system option.

File: clang/docs/ReleaseNotes.rst
Comment (ilya-biryukov) [reply clang/docs/ReleaseNotes.rst]: > Good suggestions. Applied. (Just change should to can. Since I feel should read like users can get this benefit directly. But it is not the case. They need helps from build systems).

Yeah, makse sense, "can" fits better here, we also want some follow up from the build system vendors.

> And I put this into the What's New in Clang |release|? section, which is highest except the Potentially Breaking Change section.

This seems very reasonable, I am not sure this is a "potentially breaking change" even. It does change the BMI, but that's expected with every new version of the compiler.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [reply clang/docs/StandardCPlusPlusModules.rst]: Yeah, I meant the example we already have. The current state LG, resolving this thread.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [inline clang/docs/StandardCPlusPlusModules.rst:658]: This may have been discussed somewhere that I missed, please point me to that if that's the case.

I was putting a little more thought into it and am wondering what would be a way for the build systems to take advantage of that?

When compiling something that dependends on modules, the build system will need to check if any of the (transitive) dependencies of the modules changes, right? Because any BMI that was changed in the transitive set can potentially affect the result of the current compilations.

So even if direct dependencies did not change (because of this optimization), despite a change in the deeper transitive dependency, the build system still can't reuse the result and has to rerun the compile.

We only know that the output BMI wasn't affected by any of the transitive changes **after** we finish compiling it and can compare the outputs, right?
But the build system needs to know about it **before** the compilation happens to avoid recompilations.

I suspect I'm missing something, but don't know what...

Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst:658]: > This may have been discussed somewhere that I missed, please point me to that if that's the case.
> 
> I was putting a little more thought into it and am wondering what would be a way for the build systems to take advantage of that?
> 
> When compiling something that dependends on modules, the build system will need to check if any of the (transitive) dependencies of the modules changes, right? Because any BMI that was changed in the transitive set can potentially affect the result of the current compilations.
> 
> So even if direct dependencies did not change (because of this optimization), despite a change in the deeper transitive dependency, the build system still can't reuse the result and has to rerun the compile.

Yes, and this is the reason why it is experimental.

> 
> We only know that the output BMI wasn't affected by any of the transitive changes **after** we finish compiling it and can compare the outputs, right? But the build system needs to know about it **before** the compilation happens to avoid recompilations.
> 
> I suspect I'm missing something, but don't know what...

It is not about to avoid the re-compilation for that unchanged BMI, but for other TUs that only dependent on the unchanged BMI. For example,

```
// a.cppm
export module a;
// intentional empty
```

```
// b.cppm
export module b;
import a;
// intentional empty
```

```
// c.cpp
import b;
...
```

For this example, every time `a.cppm` changes, we need to recompile `b.cppm`. But if the BMI of module `B` doesn't change at all, it should be good to not recompile `c.cpp`. Do this make sense?

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: > On "experimental": thanks for clarifying that you meant this as "experimental in the build system". I am afraid that the current wording actually does the opposite and tells that this the compiler support is "experimental" (aka "unstable", aka "don't touch this until it stabilizes").

Yeah, and I like to call it as `experimental with the build system`. Since although the users can only get it from build systems, the core abilities are provided by the compiler. Also when bug reports comes, I believe most of them need to solve in the compiler side. But this doesn't matter though : )

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/ReleaseNotes.rst
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:166]:
```suggestion
- Implemented improvements to BMIs for C++20 Modules that can reduce
  the number of rebuilds during incremental recompilation. We are seeking
  feedback from Build System authors and other interested users, especially
  when you feel Clang changes the BMI and misses an opportunity to avoid
  recompilations or causes correctness issues. See StandardCPlusPlusModules
  `StandardCPlusPlusModules <StandardCPlusPlusModules.html>`_ for more details.

```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:655]:
```suggestion
Experimental Non-Cascading Changes
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:660]:
```suggestion
if you don't want to read it all, this is helpful to reduce recompilations.
We encourage build system vendors and end users try this out and bring feedback.
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:730]:
```suggestion
It should be fine because the altered interfaces do not affect ``Use.cc`` in any way;
the changes do not cascade.
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:746]:
```suggestion
With reduced BMI, non-cascading changes can be more powerful. For example,
```

File: clang/docs/StandardCPlusPlusModules.rst
Comment (AaronBallman) [inline clang/docs/StandardCPlusPlusModules.rst:788]:
```suggestion
We should find the contents of ``B.pcm`` remains the same. In this case, the build system is
allowed to skip recompilations of TUs which solely and directly depend on module ``B``.

This only happens with a reduced BMI. With reduced BMIs, we won't record the function body
of ``int b()`` in the BMI for ``B`` so that the module ``A`` doesn't contribute to the BMI of ``B``
and we have less dependencies.
```

**Review by AaronBallman — COMMENTED** [review]
Some minor corrections for grammar.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ilya-biryukov) [reply clang/docs/StandardCPlusPlusModules.rst:658]: That's what I thought is happening, thanks for confirming my intuition.
I think it'd be great to spell this out in the documentation, because I believe this mode of operation is something that the build system vendors might need to infer from the whole text right now.

Adding something like the following should do the trick:

```
We encourage build systems to add an experimental mode that
reuses the cached BMI when **direct** dependencies did not change,
even if **transitive** dependencies did change.
```

PS My 2 cents on Bazel: I suspect this actually goes against its design. Bazel has very strong capabilities to avoid recompilations when inputs don't change, but it relies on stable hashes for all inputs (including transitive dependencies). 
At the same time I suspect that other build systems (`CMake` + `ccache` / `sccache`) are more flexible in that regard.

PPS I am not a build system expert, so take that with a grain of salt. I'm also happy to loop in Bazel folks I work with to confirm or rebut my claims about Bazel's design, if that's useful.

File: clang/docs/ReleaseNotes.rst
Comment (ChuanqiXu9) [reply clang/docs/ReleaseNotes.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst:658]: Yeah, added suggested text. 

For bazel, I think it might be fine since it should be configurable that whether not to recompile a target as far as I know.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

File: clang/docs/StandardCPlusPlusModules.rst
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Done

**Review by AaronBallman — APPROVED** [review]
LGTM but please wait for other active reviewers to sign off before landing.

File: clang/docs/StandardCPlusPlusModules.rst
Comment (mathstuf) [reply clang/docs/StandardCPlusPlusModules.rst]: Ok, so what do I do differently? AFAIK, CMake is relying on the compiler's `-MF` reporting mechanisms here.

Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: In the higher level, after offering an experimental option to enable the mode, I think what build systems to do is, when considering the recompilations, the build system don't need to consider the indirectly imported modules. And for `-MF`, as far as I know, it will collect the included headers and headers shouldn't take part in here.

For example,

```
// a.cpp
#include "a.h"
import a;
...
```

If `a.h` changes, I think we should recompile `a.cpp`.

For implementations, I don't know the implementations of cmake. But I remember you said somewhere that it might be possible to make it by a compile-and-swap action.

Comment (mathstuf) [reply clang/docs/StandardCPlusPlusModules.rst]: Ah, ok.

FWIW, I was thinking that even `a.pcm` would get reported in `-MF` for your example and elided if it didn't contribute. As for the `restat = 1` with a swap, yes, that is build-system side. I think some docs saying something like:

```md
Build systems may utilize this optimization by doing an update-if-changed operation to the BMI
that is consumed from the BMI that is output by the compiler.
```

Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Suggestion Added

**Comment by ChuanqiXu9** [comment]
I am going to land this in later of the week if no objections come. As 19.x is going to be branched.

**Pull request merged**
